Automattic/woocommerce-payments
# RPP: Handle failed transaction rate limiter

Fixes #7414

#### Changes proposed in this Pull Request

- Migrate legacy `Session_Rate_Limiter` to `SessionRateLimiter` in `src`. 
- Integrate this new class into `InitialState`.
- Add tests.
- Prepare #7734 to clean up in Jan 2024. 

## Todos

- [x] Get feedback regarding dealing with generating a new class for `SessionRateLimiter`.
- [ ] Get feedback in using `DG\BypassFinals` for testing `final` methods/classes. 









#### Testing instructions





* Use declined cards for checkout https://woo.com/document/woopayments/testing-and-troubleshooting/testing/#declined-test-cards
* After 5 times of failure, see the error message `Your payment was not processed.` in the checkout page. Tip: can change this number `5` to `2` [here](https://github.com/Automattic/woocommerce-payments/blob/f48f4fcde4ce0c481c1996e5d06d84988c63d81f/src/Internal/Payment/FailedTransactionRateLimiter.php#L29-L29) so that it's faster to see the expected result.
* Check the note section of this order, see a relevant note that the rate limiter has been enabled. 
* Put breakpoints in the new code to ensure they're run through during manual tests.

-------------------


**Post merge**
